[BUG] scipy 1.8 breaks hyppo import

#### Reproducing code example:


```
import hyppo.discrim
```

#### Error message
```
/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/hyppo/discrim/__init__.py:1: in <module>
    from .discrim_one_samp import DiscrimOneSample
/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/hyppo/discrim/discrim_one_samp.py:5: in <module>
    from ._utils import _CheckInputs
/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/hyppo/discrim/_utils.py:4: in <module>
    from ..tools import check_ndarray_xy, check_reps, contains_nan, convert_xy_float64
/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/hyppo/tools/__init__.py:1: in <module>
    from .common import *
/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/hyppo/tools/common.py:6: in <module>
    from scipy.stats.stats import _contains_nan
E   ImportError: cannot import name '_contains_nan' from 'scipy.stats.stats' (/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/scipy/stats/stats.py)
```

The function `_contains_nan()` was moved to a different file in scipy 1.8. In general, this is why it's a bad idea to rely on importing private functions.

Recommend just copy-pasting the private function into hyppo explicitly.

--- a/docs/conf.py
+++ b/docs/conf.py
@@ -25,7 +25,7 @@ sys.path.insert(0, os.path.abspath(".."))
 year = datetime.date.today().year
 project = "hyppo"
 copyright = "2018-{}, NeuroData".format(year)
-authors = u"Sambit Panda"
+authors = "Sambit Panda"
 
 # The short X.Y version
 # Find hyppo version.

--- a/hyppo/d_variate/dhsic.py
+++ b/hyppo/d_variate/dhsic.py
@@ -101,10 +101,10 @@ class dHsic(DVariateTest):
         term3 = (2 / n) * np.ones((n,))
         for kern in kerns:
             term1 = np.multiply(term1, kern)
-            term2 = (1 / n ** 2) * term2 * np.sum(kern)
+            term2 = (1 / n**2) * term2 * np.sum(kern)
             term3 = (1 / n) * np.multiply(term3, np.sum(kern, axis=1))
 
-        stat = (1 / n ** 2) * np.sum(term1) + term2 - np.sum(term3)
+        stat = (1 / n**2) * np.sum(term1) + term2 - np.sum(term3)
         self.stat = stat
 
         return stat

--- a/hyppo/independence/cca.py
+++ b/hyppo/independence/cca.py
@@ -73,8 +73,8 @@ class CCA(IndependenceTest):
 
         # if 1-d, don't calculate the svd
         if varx.size == 1 or vary.size == 1 or covar.size == 1:
-            covar = np.sum(covar ** 2)
-            stat = covar / np.sqrt(np.sum(varx ** 2) * np.sum(vary ** 2))
+            covar = np.sum(covar**2)
+            stat = covar / np.sqrt(np.sum(varx**2) * np.sum(vary**2))
         else:
             covar = np.sum(np.linalg.svd(covar, 1)[1] ** 2)
             stat = covar / np.sqrt(

--- a/hyppo/independence/dcorr.py
+++ b/hyppo/independence/dcorr.py
@@ -378,7 +378,7 @@ def _fast_1d_dcov(x, y, bias=False):  # pragma: no cover
     ) + (si[-1] - 2 * si)
 
     if bias:
-        denom = [n ** 2, n ** 3, n ** 4]
+        denom = [n**2, n**3, n**4]
     else:
         denom = [n * (n - 3), n * (n - 3) * (n - 2), n * (n - 3) * (n - 2) * (n - 1)]
 
@@ -404,7 +404,7 @@ def _dcov(distx, disty, bias=False, only_dcov=True):  # pragma: no cover
     if only_dcov:
         N = distx.shape[0]
         if bias:
-            stat = 1 / (N ** 2) * stat
+            stat = 1 / (N**2) * stat
         else:
             stat = 1 / (N * (N - 3)) * stat
 

--- a/hyppo/independence/tests/test_mgc.py
+++ b/hyppo/independence/tests/test_mgc.py
@@ -57,8 +57,8 @@ class TestMGCStat(object):
     @pytest.mark.parametrize(
         "sim, obs_stat, obs_pvalue",
         [
-            (linear, 0.97, 1 / 1000),  # test linear simulation
-        ],
+            (linear, 0.97, 1 / 1000),
+        ],  # test linear simulation
     )
     def test_rep(self, sim, obs_stat, obs_pvalue):
         x, y = sim(n=100, p=5)

--- a/hyppo/kgof/_utils.py
+++ b/hyppo/kgof/_utils.py
@@ -92,7 +92,7 @@ def fit_gaussian_draw(X, J, seed=28, reg=1e-7, eig_pow=1.0):
     evals = np.maximum(0, np.real(evals))
     assert np.all(np.isfinite(evals))
     evecs = np.real(evecs)
-    shrunk_cov = evecs.dot(np.diag(evals ** eig_pow)).dot(evecs.T) + reg * np.eye(d)
+    shrunk_cov = evecs.dot(np.diag(evals**eig_pow)).dot(evecs.T) + reg * np.eye(d)
     V = rng.multivariate_normal(mean_x, shrunk_cov, J)
     return V
 

--- a/hyppo/kgof/fssd.py
+++ b/hyppo/kgof/fssd.py
@@ -381,7 +381,7 @@ def ustat_h1_mean_variance(fea_tensor, return_variance=True, use_unbiased=True):
     Tau = np.reshape(Xi, [n, d * J])
     if use_unbiased:
         t1 = np.sum(np.mean(Tau, 0) ** 2) * (old_div(n, float(n - 1)))
-        t2 = old_div(np.sum(np.mean(Tau ** 2, 0)), float(n - 1))
+        t2 = old_div(np.sum(np.mean(Tau**2, 0)), float(n - 1))
         # stat is the mean
         stat = t1 - t2
     else:
@@ -393,7 +393,7 @@ def ustat_h1_mean_variance(fea_tensor, return_variance=True, use_unbiased=True):
     # compute the variance
     # mu: d*J vector
     mu = np.mean(Tau, 0)
-    variance = 4 * np.mean(np.dot(Tau, mu) ** 2) - 4 * np.sum(mu ** 2) ** 2
+    variance = 4 * np.mean(np.dot(Tau, mu) ** 2) - 4 * np.sum(mu**2) ** 2
     return stat, variance
 
 

--- a/hyppo/kgof/kernel.py
+++ b/hyppo/kgof/kernel.py
@@ -162,8 +162,8 @@ class KGauss(DifferentiableKernel, KSTKernel, LinearKSTKernel):
         # (n1, d1) = X.shape
         # (n2, d2) = Y.shape
         # assert d1==d2, 'Dimensions of the two inputs must be the same'
-        sumx2 = np.reshape(np.sum(X ** 2, 1), (-1, 1))
-        sumy2 = np.reshape(np.sum(Y ** 2, 1), (1, -1))
+        sumx2 = np.reshape(np.sum(X**2, 1), (-1, 1))
+        sumy2 = np.reshape(np.sum(Y**2, 1), (1, -1))
         D2 = sumx2 - 2 * np.dot(X, Y.T) + sumy2
         K = np.exp(old_div(-D2, (2.0 * self.sigma2)))
         return K
@@ -229,7 +229,7 @@ class KGauss(DifferentiableKernel, KSTKernel, LinearKSTKernel):
         assert d1 == d2, "Dimensions of the two inputs must be the same"
         d = d1
         sigma2 = self.sigma2
-        D2 = np.sum(X ** 2, 1)[:, np.newaxis] - 2 * np.dot(X, Y.T) + np.sum(Y ** 2, 1)
+        D2 = np.sum(X**2, 1)[:, np.newaxis] - 2 * np.dot(X, Y.T) + np.sum(Y**2, 1)
         K = np.exp(old_div(-D2, (2.0 * sigma2)))
         G = K / sigma2 * (d - old_div(D2, sigma2))
         return G

--- a/hyppo/ksample/energy.py
+++ b/hyppo/ksample/energy.py
@@ -115,7 +115,7 @@ class Energy(KSampleTest):
 
         # exact equivalence transformation Dcorr and Energy
         stat = (
-            _dcov(distx, disty, self.bias) * (2 * (n ** 2) * (m ** 2)) / ((n + m) ** 4)
+            _dcov(distx, disty, self.bias) * (2 * (n**2) * (m**2)) / ((n + m) ** 4)
         )
         self.stat = stat
 

--- a/hyppo/ksample/mean_embedding.py
+++ b/hyppo/ksample/mean_embedding.py
@@ -166,7 +166,7 @@ def mean_embed_distance(difference, num_randfeatures):
     mu = np.mean(difference, 0)
 
     if num_randfeatures == 1:
-        stat = float(num_samples * mu ** 2) / float(sigma)
+        stat = float(num_samples * mu**2) / float(sigma)
     else:
         stat = num_samples * mu.dot(np.linalg.solve(sigma, np.transpose(mu)))
 

--- a/hyppo/ksample/mmd.py
+++ b/hyppo/ksample/mmd.py
@@ -112,7 +112,7 @@ class MMD(KSampleTest):
                 x,
                 y,
             )
-            * (2 * (n ** 2) * (m ** 2))
+            * (2 * (n**2) * (m**2))
             / ((n + m) ** 4)
         )
         self.stat = stat

--- a/hyppo/time_series/tests/test_mgcx.py
+++ b/hyppo/time_series/tests/test_mgcx.py
@@ -44,7 +44,7 @@ class TestMGCXStat:
     def test_optimal_scale_nonlinear(self):
         n = 7
         x = np.arange(n)
-        y = x ** 2
+        y = x**2
         mgcx_dict = MGCX().test(x, y, reps=100)[2]
         opt_scale = mgcx_dict["opt_scale"]
         assert_array_less(opt_scale[0], n)
@@ -64,7 +64,7 @@ class TestMGCXStat:
     def test_rep_nonlinear(self):
         n = 7
         x = np.arange(n)
-        y = x ** 2
+        y = x**2
         mgcx_dict = MGCX().test(x, y, reps=100, random_state=2)[2]
         opt_scale = mgcx_dict["opt_scale"]
         mgcx_dict2 = MGCX().test(x, y, reps=100, random_state=2)[2]

--- a/hyppo/tools/common.py
+++ b/hyppo/tools/common.py
@@ -3,15 +3,39 @@ import warnings
 import numpy as np
 from joblib import Parallel, delayed
 from scipy.stats.distributions import chi2
-from scipy.stats.stats import _contains_nan
 from sklearn.metrics import pairwise_distances
 from sklearn.metrics.pairwise import pairwise_kernels
 from sklearn.utils import check_random_state
 
 
-def contains_nan(a):  # from scipy
-    """Check if inputs contains NaNs"""
-    return _contains_nan(a, nan_policy="raise")
+# Explicitly copying private function from scipy 1.7.3
+# Modified to only use nan_policy 'raise'
+# REF: https://github.com/scipy/scipy/blob/59e6539cf80dc04b16b0f0ab52343381f0a7a2fa/scipy/stats/stats.py#L79
+def contains_nan(a):
+    nan_policy = 'raise'
+    try:
+        # Calling np.sum to avoid creating a huge array into memory
+        # e.g. np.isnan(a).any()
+        with np.errstate(invalid='ignore'):
+            contains_nan_var = np.isnan(np.sum(a))
+    except TypeError:
+        # This can happen when attempting to sum things which are not
+        # numbers (e.g. as in the function `mode`). Try an alternative method:
+        try:
+            contains_nan_var = np.nan in set(a.ravel())
+        except TypeError:
+            # Don't know what to do. Fall back to omitting nan values and
+            # issue a warning.
+            contains_nan_var = False
+            nan_policy = 'omit'
+            warnings.warn("The input array could not be properly "
+                          "checked for nan values. nan values "
+                          "will be ignored.", RuntimeWarning)
+
+    if contains_nan_var:
+        raise ValueError("The input contains nan values")
+
+    return contains_nan_var, nan_policy
 
 
 def check_ndarray_xy(x, y):

--- a/hyppo/tools/common.py
+++ b/hyppo/tools/common.py
@@ -175,7 +175,7 @@ def compute_kern(x, y, metric="gaussian", workers=1, **kwargs):
             )
             # prevents division by zero when used on label vectors
             med = med if med else 1
-            kwargs["gamma"] = 1.0 / (2 * (med ** 2))
+            kwargs["gamma"] = 1.0 / (2 * (med**2))
         metric = "rbf"
     if callable(metric):
         simx = metric(x, **kwargs)
@@ -260,7 +260,7 @@ def multi_compute_kern(*args, metric="gaussian", workers=1, **kwargs):
             )
             # prevents division by zero when used on label vectors
             med = med if med else 1
-            kwargs["gamma"] = 1.0 / (2 * (med ** 2))
+            kwargs["gamma"] = 1.0 / (2 * (med**2))
         metric = "rbf"
     if callable(metric):
         sim_mats = []

--- a/hyppo/tools/common.py
+++ b/hyppo/tools/common.py
@@ -12,11 +12,11 @@ from sklearn.utils import check_random_state
 # Modified to only use nan_policy 'raise'
 # REF: https://github.com/scipy/scipy/blob/59e6539cf80dc04b16b0f0ab52343381f0a7a2fa/scipy/stats/stats.py#L79
 def contains_nan(a):
-    nan_policy = 'raise'
+    nan_policy = "raise"
     try:
         # Calling np.sum to avoid creating a huge array into memory
         # e.g. np.isnan(a).any()
-        with np.errstate(invalid='ignore'):
+        with np.errstate(invalid="ignore"):
             contains_nan_var = np.isnan(np.sum(a))
     except TypeError:
         # This can happen when attempting to sum things which are not
@@ -27,10 +27,13 @@ def contains_nan(a):
             # Don't know what to do. Fall back to omitting nan values and
             # issue a warning.
             contains_nan_var = False
-            nan_policy = 'omit'
-            warnings.warn("The input array could not be properly "
-                          "checked for nan values. nan values "
-                          "will be ignored.", RuntimeWarning)
+            nan_policy = "omit"
+            warnings.warn(
+                "The input array could not be properly "
+                "checked for nan values. nan values "
+                "will be ignored.",
+                RuntimeWarning,
+            )
 
     if contains_nan_var:
         raise ValueError("The input contains nan values")

--- a/hyppo/tools/indep_sim.py
+++ b/hyppo/tools/indep_sim.py
@@ -195,9 +195,9 @@ def cubic(n, p, noise=False, low=-1, high=1, cubs=[-12, 48, 128], scale=1 / 3):
 
     x_coeffs = x @ coeffs - scale
     y = (
-        cubs[2] * x_coeffs ** 3
-        + cubs[1] * x_coeffs ** 2
-        + cubs[0] * x_coeffs ** 3
+        cubs[2] * x_coeffs**3
+        + cubs[1] * x_coeffs**2
+        + cubs[0] * x_coeffs**3
         + 80 * noise * eps
     )
 
@@ -354,7 +354,7 @@ def quadratic(n, p, noise=False, low=-1, high=1):
     eps = _calc_eps(n)
 
     x_coeffs = x @ coeffs
-    y = x_coeffs ** 2 + 0.5 * noise * eps
+    y = x_coeffs**2 + 0.5 * noise * eps
 
     return x, y
 
@@ -407,7 +407,7 @@ def w_shaped(n, p, noise=False, low=-1, high=1):
 
     x_coeffs = x @ coeffs
     u_coeffs = u @ coeffs
-    y = 4 * ((x_coeffs ** 2 - 0.5) ** 2 + u_coeffs / 500) + 0.5 * noise * eps
+    y = 4 * ((x_coeffs**2 - 0.5) ** 2 + u_coeffs / 500) + 0.5 * noise * eps
 
     return x, y
 
@@ -565,7 +565,7 @@ def logarithmic(n, p, noise=False):
     x = np.random.multivariate_normal(np.zeros(p), sig, size=n)
     eps = _calc_eps(n)
 
-    y = np.log(x ** 2) + 3 * noise * eps
+    y = np.log(x**2) + 3 * noise * eps
 
     return x, y
 
@@ -826,7 +826,7 @@ def two_parabolas(n, p, noise=False, low=-1, high=1, prob=0.5):
     rand_noise = _random_uniform(n, p, low=0, high=1)
 
     x_coeffs = x @ coeffs
-    y = (x_coeffs ** 2 + 2 * noise * rand_noise) * (u - 0.5)
+    y = (x_coeffs**2 + 2 * noise * rand_noise) * (u - 0.5)
 
     return x, y
 

--- a/hyppo/tools/ksample_sim.py
+++ b/hyppo/tools/ksample_sim.py
@@ -36,7 +36,7 @@ def _2samp_rotate(sim, x, y, p, degree=90, pow_type="samp"):
         ]:
             for i in range(rot_shape):
                 mat = np.random.normal(size=(rot_shape, 1))
-                mat = mat / np.sqrt(np.sum(mat ** 2))
+                mat = mat / np.sqrt(np.sum(mat**2))
                 if i == 0:
                     rot = mat
                 else:

--- a/hyppo/tools/tests/test_common.py
+++ b/hyppo/tools/tests/test_common.py
@@ -185,7 +185,7 @@ class TestHelper:
                 l1, (n - 1, n + 1), (l1.itemsize * (n + 1), l1.itemsize)
             )[:, 1:]
         )
-        gamma = 1.0 / (2 * (med ** 2))
+        gamma = 1.0 / (2 * (med**2))
 
         kernx = pairwise_kernels(x, x, metric="rbf", gamma=gamma)
         kerny = pairwise_kernels(y, y, metric="rbf", gamma=gamma)
@@ -230,7 +230,7 @@ class TestHelper:
                 l1, (n - 1, n + 1), (l1.itemsize * (n + 1), l1.itemsize)
             )[:, 1:]
         )
-        gamma = 1.0 / (2 * (med ** 2))
+        gamma = 1.0 / (2 * (med**2))
 
         kernx = pairwise_kernels(x, x, metric="rbf", gamma=gamma)
         kerny = pairwise_kernels(y, y, metric="rbf", gamma=gamma)

--- a/tutorials/independence.py
+++ b/tutorials/independence.py
@@ -97,8 +97,8 @@ t_fast_hhg = timeit.Timer(stmt="HHG().test(x, y, auto=True)", setup=setup_code)
 hhg_time = np.array(t_hhg.timeit(number=1))  # original HHG
 fast_hhg_time = np.array(t_fast_hhg.timeit(number=5)) / 5  # fast HHG
 
-print(u"Original HHG time: {0:.3g}s".format(hhg_time))
-print(u"Fast HHG time: {0:.3g}s".format(fast_hhg_time))
+print("Original HHG time: {0:.3g}s".format(hhg_time))
+print("Fast HHG time: {0:.3g}s".format(fast_hhg_time))
 
 ########################################################################################
 # ------------
@@ -147,10 +147,10 @@ chisq_time = np.array(t_chisq.timeit(number=1000)) / 1000  # fast Dcorr
 fast_perm_time = np.array(t_fast_perm.timeit(number=1))  # permutation Dcorr
 fast_chisq_time = np.array(t_fast_chisq.timeit(number=1000)) / 1000  # fast Dcorr
 
-print(u"Permutation time: {0:.3g}s".format(perm_time))
-print(u"Fast time (chi-square): {0:.3g}s".format(chisq_time))
-print(u"Permutation time (fast statistic): {0:.3g}s".format(fast_perm_time))
-print(u"Fast time (fast statistic chi-square): {0:.3g}s".format(fast_chisq_time))
+print("Permutation time: {0:.3g}s".format(perm_time))
+print("Fast time (chi-square): {0:.3g}s".format(chisq_time))
+print("Permutation time (fast statistic): {0:.3g}s".format(fast_perm_time))
+print("Fast time (fast statistic chi-square): {0:.3g}s".format(fast_chisq_time))
 
 ########################################################################################
 # Look at the time increases when using the fast test!

--- a/tutorials/kgof.py
+++ b/tutorials/kgof.py
@@ -61,24 +61,26 @@ import matplotlib.pyplot as plt
 import autograd.numpy as np
 from numpy.random import default_rng
 
+
 def isogauss_log_den(X):
-  mean = np.zeros(2)
-  variance = 1
-  unden = -np.sum((X - mean)**2, 1) / (2.0 * variance)
-  return unden
+    mean = np.zeros(2)
+    variance = 1
+    unden = -np.sum((X - mean) ** 2, 1) / (2.0 * variance)
+    return unden
+
 
 ########################################################################################
 # This function computes the log of an unnormalized density. This works fine as this test
-# only requires a :math:`\nabla_{\mathbf{x}} \log p(\mathbf{x})` which does not depend on 
-# the normalizer. The gradient :math:`\nabla_{\mathbf{x}} \log p(\mathbf{x})` will be 
-# automatically computed by autograd. In this kgof package, a model :math:`p` can be 
+# only requires a :math:`\nabla_{\mathbf{x}} \log p(\mathbf{x})` which does not depend on
+# the normalizer. The gradient :math:`\nabla_{\mathbf{x}} \log p(\mathbf{x})` will be
+# automatically computed by autograd. In this kgof package, a model :math:`p` can be
 # specified by implementing the class :class:`hyppo.kgof.density.UnnormalizedDensity`. Implementing
-# this directly is a bit tedious, however. Construct an :class:`UnnormalizedDensity` which will 
+# this directly is a bit tedious, however. Construct an :class:`UnnormalizedDensity` which will
 # represent a Gaussian model. All the implemented goodness-of-fit tests take this object as input.
 
 # Next, draw a sample from q.
 # Drawing n points from q
-m = 1 # If m = 0, p = q and H_0 is true
+m = 1  # If m = 0, p = q and H_0 is true
 
 seed = 4
 rng = default_rng(seed)
@@ -86,18 +88,18 @@ n = 400
 X = rng.standard_normal(size=(n, 2)) + np.array([m, 0])
 
 # Plot the data from q
-plt.plot(X[0], X[1], 'ko', label='Data from $q$')
+plt.plot(X[0], X[1], "ko", label="Data from $q$")
 plt.legend()
 
 ########################################################################################
-# All the implemented tests take the data in the form of a :class:`hyppo.kgof.data.Data` object. 
+# All the implemented tests take the data in the form of a :class:`hyppo.kgof.data.Data` object.
 # This is just an encapsulation of the sample :math:`X`. To construct :class:`data.Data` do the following:
 
 dat = Data(X)
 
 ########################################################################################
-# Conduct an FSSD kernel goodness-of-fit test for an isotropic normal density with mean 
-# 0 and variance 1. 
+# Conduct an FSSD kernel goodness-of-fit test for an isotropic normal density with mean
+# 0 and variance 1.
 
 mean = 0
 variance = 1
@@ -114,4 +116,3 @@ fssd = FSSD(isonorm, k, V, null_sim=null_sim, alpha=0.01)
 
 tresult = fssd.test(dat, return_simulated_stats=True)
 tresult
-

